File: Go/123. Best Time to Buy and Sell Stock III.go
func maxProfit(prices []int) int {
	if (len(prices) == 0) {
		return 0
	}
	return helper(prices, 2)
}

func helper(prices []int, k int) int {
	var dp [][]int
	dp = make([][]int, k + 1, k + 1)
	for i := 0; i <= k; i++ {
		dp[i] = make([]int, len(prices), len(prices))
	}
	for i := 1; i <= k; i++ {
		max := dp[i - 1][0] - prices[0]
		for j := 1; j < len(prices); j++ {
			dp[i][j] = dp[i][j - 1]
			if max < dp[i - 1][j] - prices[j] {
				max = dp[i - 1][j] - prices[j]
			}
			if dp[i][j] < max + prices[j] {
				dp[i][j] = max + prices[j]
			}
		}
	}
	return dp[k][len(prices) - 1]
}

File: C++/123. Best Time to Buy and Sell Stock III.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    int maxProfit(std::vector<int>& prices) {
        if (prices.size() <= 1)
            return 0;
        int k = 2;
        int N = int(prices.size());
        std::vector<std::vector<int>> dp (k + 1, std::vector<int>(N, 0));
        for (int i = 1; i <= k; i++) {
            int Max = dp[i - 1][0] - prices[0];
            for (int j = 1; j < N; j++) {
                Max = std::max(Max, dp[i - 1][j] - prices[j]);
                dp[i][j] = std::max(dp[i][j - 1], Max + prices[j]);
            }
        }
        return dp[k][N - 1];
    }
};